compiler/src/dotty/tools/debug/InsertExpression.scala

```diff
@@ -0,0 +1,238 @@
+            expressionInserted = false
+            cpy.PackageDef(transformed)(
+              transformed.pid,
+              transformed.stats ++ expressionClass.map(_.withSpan(tree.span))
+            )
+          else transformed
+        case tree @ DefDef(name, paramss, tpt, rhs) if rhs != EmptyTree && isOnBreakpoint(tree) =>
+          cpy.DefDef(tree)(name, paramss, tpt, mkExprBlock(expression, tree.rhs))
+        case tree @ Match(selector, caseDefs) if isOnBreakpoint(tree) || caseDefs.exists(isOnBreakpoint) =>
+          // the expression is on the match or a case of the match
+          // if it is on the case of the match the program could pause on the pattern, the guard or the body
+          // we assume it pauses on the pattern because that is the first instruction
+          // in that case we cannot compile the expression val in the pattern, but we can compile it in the selector
+          cpy.Match(tree)(mkExprBlock(expression, selector), caseDefs)
+        case tree @ ValDef(name, tpt, _) if isOnBreakpoint(tree) =>
+          cpy.ValDef(tree)(name, tpt, mkExprBlock(expression, tree.rhs))
+        case tree @ PatDef(mods, pat, tpt, rhs) if isOnBreakpoint(tree) =>
+          PatDef(mods, pat, tpt, mkExprBlock(expression, rhs))
+        case tree: (Ident | Select | GenericApply | Literal | This | New | InterpolatedString | OpTree | Tuple |
+              Assign | Block) if isOnBreakpoint(tree) =>
+          mkExprBlock(expression, tree)
+
+        // for loop: we insert the expression on the first enumeration
+        case tree @ ForYield(enums, rhs) if isOnBreakpoint(tree) =>
+          ForYield(transform(enums.head) :: enums.tail, rhs)
+        case tree @ ForDo(enums, rhs) if isOnBreakpoint(tree) =>
+          ForDo(transform(enums.head) :: enums.tail, rhs)
+
+        // generator of for loop: we insert the expression on the rhs
+        case tree @ GenFrom(pat, rhs, checkMode) if isOnBreakpoint(tree) =>
+          GenFrom(pat, mkExprBlock(expression, rhs), checkMode)
+        case tree @ GenAlias(pat, rhs) if isOnBreakpoint(tree) =>
+          GenAlias(pat, mkExprBlock(expression, rhs))
+
+        case tree => super.transform(tree)
+
+  private def parseExpression(using Context): Tree =
+    val prefix =
+      s"""|object Expression:
+          |  {
+          |    """.stripMargin
+    // don't use stripMargin on wrappedExpression because expression can contain a line starting with `  |`
+    val wrappedExpression = prefix + config.expression + "\n  }\n"
+    val expressionFile = SourceFile.virtual("<expression>", config.expression)
+    val contentBytes = wrappedExpression.getBytes(StandardCharsets.UTF_8)
+    val wrappedExpressionFile =
+      new VirtualFile("<wrapped-expression>", contentBytes)
+    val sourceFile =
+      new SourceFile(wrappedExpressionFile, wrappedExpression.toArray):
+        override def start: Int =
+          // prefix.size depends on the OS
+          -prefix.size
```

Comment:
Really!? If this source file is encoded with UNIX line breaks, then `prefix` should always contain UNIX line breaks, doesn't it?

Reply:
removed confusing comment